Should shadow indices reuse allocation ids when promoting a replica to primary

Shadow replica to primary promotion is special in that that it requires the `ShadowIndexShard` of the replica to be replaced with a new `IndexShard` instance. At the moment this is achieved be giving the shard a new allocation id, triggering a cleanup in `IndicesClusterStateService`. This is the easiest solution from the `IndicesClusterStateService` perspective but it have the down side that it cause the in sync replica list to be out of date / empty and may cause problems if the cluster is shut down before the new primary is started. We should decide whether we want to align this with how normal shard work (reuse of allocation ids) or come up with another solutions.

This was all discovered while working on #20637 and adding extra validation logic [here](https://github.com/elastic/elasticsearch/pull/20637/commits/3b23135cefd6a49c8942d93c929a021255d0aef4#diff-814ec451d0e00936e3a5f9cb6811e5bdR143)

Another things that play here is that currently the promotion of a replica to primary also removes the old aid from the in-sync replicas list, causing it to only contain other replicas. If those replicas fail during the initialization of the newly promoted replica, we trigger the assertion we have [here](https://github.com/elastic/elasticsearch/blob/1e6a882ab9174e6c00954264f859086a16d1f8b3/core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java#L199) , causing the follwing failure:

```
  > Throwable #1: java.lang.AssertionError: in-sync set became empty but active primary wasn't failed: [WU7AocjrSB-sG0IOflGj8w]
   > 	at __randomizedtesting.SeedInfo.seed([B37B65076A2E7CB1:CBFC6CD43515AAF8]:0)
   > 	at org.elasticsearch.cluster.routing.allocation.IndexMetaDataUpdater.updateInSyncAllocations(IndexMetaDataUpdater.java:199)
   > 	at org.elasticsearch.cluster.routing.allocation.IndexMetaDataUpdater.applyChanges(IndexMetaDataUpdater.java:122)
```

See https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+dockeralpine-periodic/58

--- a/core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java
+++ b/core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java
@@ -390,7 +390,7 @@ public final class ShardRouting implements Writeable, ToXContent {
         assert primary : this;
         return new ShardRouting(shardId, currentNodeId, null, primary, ShardRoutingState.INITIALIZING,
             StoreRecoverySource.EXISTING_STORE_INSTANCE, new UnassignedInfo(UnassignedInfo.Reason.REINITIALIZED, null),
-            AllocationId.newInitializing(), UNAVAILABLE_EXPECTED_SHARD_SIZE);
+            allocationId, UNAVAILABLE_EXPECTED_SHARD_SIZE);
     }
 
     /**

--- a/core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java
+++ b/core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java
@@ -93,10 +93,6 @@ public class IndexMetaDataUpdater extends RoutingChangesObserver.AbstractRouting
         removeAllocationId(removedRelocationSource);
     }
 
-    @Override
-    public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard) {
-    }
-
     /**
      * Updates the current {@link MetaData} based on the changes of this RoutingChangesObserver. Specifically
      * we update {@link IndexMetaData#getInSyncAllocationIds()} and {@link IndexMetaData#primaryTerm(int)} based on

--- a/core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java
+++ b/core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java
@@ -95,7 +95,6 @@ public class IndexMetaDataUpdater extends RoutingChangesObserver.AbstractRouting
 
     @Override
     public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard) {
-        removeAllocationId(startedPrimaryShard);
     }
 
     /**

--- a/core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdTests.java
+++ b/core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdTests.java
@@ -122,7 +122,7 @@ public class AllocationIdTests extends ESTestCase {
         shard = shard.reinitializePrimaryShard();
         assertThat(shard.allocationId().getId(), notNullValue());
         assertThat(shard.allocationId().getRelocationId(), nullValue());
-        assertThat(shard.allocationId().getId(), not(equalTo(allocationId.getId())));
+        assertThat(shard.allocationId().getId(), equalTo(allocationId.getId()));
     }
 
     public void testSerialization() throws IOException {

